### ISSUE TITLE
Raise an Exception if the spike times given to SpikeSourceArray are not ordered

### ISSUE THREAD

**Issue by ruthvik92** [NeuralEnsemble/PyNN#511]
`Warning in VecStim: spike time (1193.95 ms) before current time (1363.69 ms)`

Hi,
I am using numpy array(list of lists) for spike times for 729 neurons. I'm using SpikeSourceArray of PyNN.  I couldn't understand the warning, please elaborate it further. Am I losing any spikes here?

**Comment by appukuttan-shailesh** [comment]
The warning arises from pyNN.NEURON's vecstim.mod. Its documentation states that:

> The mechanism checks at intervals `ping` whether a new vector has been provided
> using the play() procedure and if so resets its pointer to the first element
> in the new vector. Note that any spikes remaining in the first vector will be
> lost. Any spiketimes in the new vector that are earlier than the current time
> are ignored.

The warning that you are receiving is likely owing to that last line: "Any spiketimes in the new vector that are earlier than the current time are ignored." Minimal code to reproduce this scenario would be useful to pinpoint the cause.

**Comment by ruthvik92** [comment]
My SpikeSourceArray was not sorted, I think that's the reason i received this error. Thank you.

**Comment by apdavison** [comment]
Please could you paste the error traceback? We should try to provide a more informative error message, e.g. "This may be because your spike times are not sorted".

(For information, PyNN doesn't sort the spike times for you for the sake of efficiency, since most often the times are already sorted).

**Comment by ruthvik92** [comment]
Sorry, It's a warning not an error. There was not anything else in the Warning other than the heading of this thread.

**Comment by apdavison** [comment]
I checked this with the NEST backend, and it raises an Exception:
```python
NESTError: BadProperty in SetDefaults_l_D: Setting status of model 'spike_generator': Spike times must be sorted in non-descending order.
```
With the Brian backend, there is neither warning nor error.

On balance, I think we should have PyNN raise an exception consistently across backends. I suggest implementing this separately for each backend, rather than in `common`, to avoid duplicating the checks (e.g. in NEST we just have to catch the NESTError and re-raise a PyNN-specific exception, whereas in Brian we will have to do the check ourselves).

I am re-naming the issue accordingly (old name was "Warning in VecStim: spike time (1193.95 ms) before current time (1363.69 ms)")

**Comment by ruthvik92** [comment]
Now a days I am using **Neuron**( forming synapses in much faster). I think **Neuron** also has no Warning mechanism for this issue.

**Comment by appukuttan-shailesh** [comment]
The proposed fix (#515) resolves this for Brian, NEST and NEURON with identical warnings.

**Comment by apdavison** [comment]
Fixed in #515
